### PR TITLE
Feature/etcd lookup/sterling

### PR DESCRIPTION
**Pull request by sdreyer**

### CODE REVIEW

**Review by Sami1309 — COMMENTED** [review]
Looks very good imo

File: ff/serving/metadata/etcd.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+	response := make([][]byte, resp.Count)
+	for i, res := range resp.Kvs {
+		response[i] = res.Value
+	}
+	return response, nil
+}
+
+//Returns number of keys that match key prefix
+//See GetWithPrefix for more details on prefix
+func (config EtcdConfig) GetCountWithPrefix(key string) (int64, error) {
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go:113]: good point

**Review by simba-git — CHANGES_REQUESTED** [review]
Really excited to get this in, few minor changes

File: ff/serving/go.mod [hunk 2]
```diff
@@ -20,12 +20,18 @@ require (
 require (
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/go.mod:20]: Removed it. Should add a gitignore on that file

File: ff/serving/metadata/etcd.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+//Takes a populated ETCD storage struct and a resource
+//Checks to make sure the given ETCD Storage Object contains a Resource, not job
+//Deserializes Resource value into the provided Resource object
+func (config EtcdConfig) ParseResource(res EtcdStorage, resType Resource) (Resource, error) {
+	if res.StorageType != RESOURCE {
+		return nil, fmt.Errorf("payload is not resource type")
+	}
+
+	if !resType.Proto().ProtoReflect().IsValid() {
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go:147]: Yeah, was getting null pointer exceptions when testing it with empty resources

File: ff/serving/metrics/test/metrics_test.go [hunk 4]
```diff
@@ -2,45 +2,45 @@ package main
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metrics/test/metrics_test.go:2]: Can revert it. Just got caught in gofmt

File: ff/serving/metadata/metadata.go [hunk 5]
```diff
@@ -1040,6 +1071,17 @@ func (serv *MetadataServer) genericCreate(ctx context.Context, res Resource, ini
 	return &pb.Empty{}, nil
 }
 
+//Stores new values for resource after modified by notify
+func WrapNotify(lookup ResourceLookup, op operation, newRes Resource, res Resource) error {
+	if err := res.Notify(lookup, op, newRes); err != nil {
+		return err
+	}
+	if err := lookup.Set(res.ID(), res); err != nil {
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/metadata.go]: gotcha

File: ff/serving/metadata/metadata.go [hunk 6]
```diff
@@ -801,12 +808,36 @@ func NewMetadataServer(server *Config) (*MetadataServer, error) {
...
 		lookup: lookup,
 		Logger: server.Logger,
+		etcd:   server.ETCD,
 	}, nil
 }
 
 type Config struct {
 	Logger          *zap.SugaredLogger
 	TypeSenseParams *search.TypeSenseParams
+	StorageProvider StorageProvider
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/metadata.go:841]: you go it

File: ff/serving/metrics/test/metrics_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metrics/test/metrics_test.go]: Reverted

File: ff/serving/metadata/etcd.go [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+type StorageType string
+
+const (
+	RESOURCE StorageType = "Resource"
+	JOB                  = "Job"
+)
+
+//Configuration For ETCD Cluster
+type EtcdConfig struct {
+	Host string //localhost
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go]: Removed

File: ff/serving/metadata/etcd.go [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+		break
+	default:
+		return nil, fmt.Errorf("Invalid Type\n")
+	}
+	return resource, nil
+}
+
+//Serializes the entire ETCD Storage Object to be put into ETCD
+func (lookup etcdResourceLookup) serializeResource(res Resource) ([]byte, error) {
+	p, _ := proto.Marshal(res.Proto())
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go]: Likely was testing something and forgot to put it back. Readded

File: ff/serving/metadata/etcd.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+	msg := EtcdStorage{
+		ResourceType: res.ID().Type,
+		Message:      p,
+		StorageType:  RESOURCE,
+	}
+	strmsg, err := json.Marshal(msg)
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go]: Changed Name

File: ff/serving/metadata/etcd.go [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+//Can be implemented to unmarshal EtcdStorage Object into format used by Jobs
+//Like ParseResource Above
+func (config EtcdConfig) ParseJob(res EtcdStorage) ([]byte, error) {
+	return nil, nil
+}
+
+//Returns an empty Resource Object of the given type to unmarshal etcd value into
+func (lookup etcdResourceLookup) findResourceType(t ResourceType) (Resource, error) {
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply ff/serving/metadata/etcd.go]: Renamed to createEmptyResource for clarity

**Comment by sdreyer** [comment]
Need to fix that coverage test. Actual test coverage is ~80%

**Review by simba-git — DISMISSED** [review]

**Pull request merged**